src/mds/Server.cc

```diff
@@ -353,55 +351,90 @@ void Server::handle_client_session(MClientSession *m)
       return;
     }
 
-    blacklisted = mds->objecter->with_osdmap(
-        [session](const OSDMap &osd_map) -> bool {
-          return osd_map.is_blacklisted(session->info.inst.addr);
-        });
+    {
+      auto send_reject_message = [this, session](std::string_view err_str) {
+	MClientSession *m = new MClientSession(CEPH_SESSION_REJECT);
+	if (session->info.has_feature(CEPHFS_FEATURE_MIMIC))
+	  m->metadata["error_string"] = err_str;
+	mds->send_message_client(m, session);
+      };
+
+      bool blacklisted = mds->objecter->with_osdmap(
+	  [session](const OSDMap &osd_map) -> bool {
+	    return osd_map.is_blacklisted(session->info.inst.addr);
+	  });
+
+      if (blacklisted) {
+	dout(10) << "rejecting blacklisted client " << session->info.inst.addr << dendl;
+	send_reject_message("blacklisted");
+	session->clear();
+	break;
+      }
 
-    if (blacklisted) {
-      dout(10) << "rejecting blacklisted client " << session->info.inst.addr << dendl;
-      mds->send_message_client(new MClientSession(CEPH_SESSION_REJECT), session);
-      m->put();
-      return;
-    }
+      client_metadata_t client_metadata(std::move(m->metadata),
+					std::move(m->supported_features));
+      if (client_metadata.features.empty())
+	infer_supported_features(session, client_metadata);
 
-    session->set_client_metadata(m->client_meta);
-    dout(20) << __func__ << " CEPH_SESSION_REQUEST_OPEN "
-      << session->info.client_metadata.size() << " metadata entries:" << dendl;
-    for (map<string, string>::iterator i = session->info.client_metadata.begin();
-        i != session->info.client_metadata.end(); ++i) {
-      dout(20) << "  " << i->first << ": " << i->second << dendl;
-    }
-
-    // Special case for the 'root' metadata path; validate that the claimed
-    // root is actually within the caps of the session
-    if (session->info.client_metadata.count("root")) {
-      const auto claimed_root = session->info.client_metadata.at("root");
-      // claimed_root has a leading "/" which we strip before passing
-      // into caps check
-      if (claimed_root.empty() || claimed_root[0] != '/' ||
-          !session->auth_caps.path_capable(claimed_root.substr(1))) {
-        derr << __func__ << " forbidden path claimed as mount root: "
-             << claimed_root << " by " << m->get_source() << dendl;
-        // Tell the client we're rejecting their open
-        mds->send_message_client(new MClientSession(CEPH_SESSION_REJECT), session);
-        mds->clog->warn() << "client session with invalid root '" <<
-          claimed_root << "' denied (" << session->info.inst << ")";
```

Comment:
Changing this message caused a failure: http://pulpito.ceph.com/pdonnell-2018-07-12_18:01:13-fs-wip-pdonnell-testing-20180712.152150-distro-basic-smithi/2771850/

New message doesn't match the format the test expects:

````
2018-07-13T02:22:45.033 INFO:teuthology.orchestra.run.smithi158.stdout:2018-07-13 02:22:28.722697 mds.a [WRN] client session with non-allowable root '/baz' denied (client.5050 172.21.15.158:0/2635692791)
````